File: Pattern/Tree/Validation/Min_DepthOf_BT/Min_Depth_Of_BT.cpp

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
//Iterative approach using BFS
class Solution {
public:
    int minDepth(TreeNode* root) {
        TreeNode* k;
        int cnt = 1;
       queue<TreeNode*> q;
        if (root == NULL) {
            return 0;
        }
      
        q.push(root);
        while (!q.empty()) {
            int n = q.size();
            while (n--) {

                TreeNode* k = q.front();
                q.pop();
                if (k->left == NULL && k->right == NULL) {
                    return cnt;
                }
                if (k->left != NULL) {
                    q.push(k->left);
                }
                if (k->right != NULL) {
                    q.push(k->right);
                }
            }
            cnt++;
        }
        return cnt;
    }
};


//Recursive approach

File: Pattern/Tree/Validation/Min_DepthOf_BT/MinDepthOf_BT.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int minDepth(TreeNode root) {
        

        int cnt = 1;
    Queue<TreeNode>q =new LinkedList<>();
        if (root == null) {
            return 0;
        }
      
        q.add(root);
        while (!q.isEmpty()) {
            int n = q.size();
            while (n>0) {

                TreeNode k = q.peek();
                q.remove();
                if (k.left == null && k.right == null) {
                    return cnt;
                }
                if (k.left != null) {
                    q.add(k.left);
                }
                if (k.right != null) {
                    q.add(k.right);
                }
                n--;
            }
            cnt++;
        }
        return cnt;
     

    }
}
